File: Leetcode/Binary Search Tree/1382. Balance a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        node_values = []
        
        def traverseBST(node: TreeNode) -> None:
            if node is None: return
            
            traverseBST(node.left)
            
            node_values.append(node.val)
            
            traverseBST(node.right)
        
        
        def constructBST(i: int, j: int) -> TreeNode:
            if i > j: return None
            
            mid = (i + j) // 2
            
            node_new = TreeNode(node_values[mid])
            
            node_new.left = constructBST(i, mid-1)
            node_new.right = constructBST(mid+1, j)

            return node_new
        
          
        traverseBST(root)
        
        return constructBST(0, len(node_values)-1)

File: Leetcode/Practice/Binary Search Tree/1382. Balance a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector <int> v;
    TreeNode* balanceBST(TreeNode* root) {
        unbalancedBSTtoArray(root);

        sort(v.begin(), v.end());

        return arrayToBalancedBST(0, v.size()-1);
    }

    void unbalancedBSTtoArray(TreeNode *cur) {
        if(!cur) return;

        v.push_back(cur->val);

        unbalancedBSTtoArray(cur->left);
        unbalancedBSTtoArray(cur->right);
    }

     TreeNode *arrayToBalancedBST(int st, int en) {
        if(st > en) return NULL;

        int mid = (st + en) / 2;

        TreeNode *node = new TreeNode(v[mid]);

        node->left = arrayToBalancedBST(st, mid-1);
        node->right = arrayToBalancedBST(mid+1, en);

        return node;
    }
};
